### ISSUE TITLE
SEC-45: Forward request parameters to OIDC authorization server as authentication parameter

### ISSUE THREAD

**Issue by ayoho** [OpenLiberty/open-liberty#6599]
1. Update `<openidConnectClient >` and `<oidcLogin>` to have configuration attribute `forwardAuthzParameter`.

    Example:
    ```XML
    forwardAuthzParameter="login_hint, ui_locales"
    ````

1. Add `forwardAuthzParameter` to metatype as "internal" for now until WAD is approved.

1. If the configured attribute contains one or more of the following reserved parameters, we should silently exclude them and issue a warning message:
    - redirect_uri
    - client_id
    - response_type
    - nonce
    - state
    - scope

1. Update runtime to extract the configured parameters from `HttpServletRequest` objects and append them to Authorization Request URL. For example, if the OIDC client has configured

    ```XML
    forwardAuthzParameter="login_hint, ui_locales"
    ```

    and

    ```Java
    request.getParameter("login_hint") = tom@example.com
    request.getParameter("ui_locales") = null
    ```

    The authorization request URL should be like this

    ```
    response_type=code
        &client_id=s6BhdRkqt3
        &scope=openid%20profile
        &state=af0ifjsldkj
        &redirect_uri=https%3A%2F%2Fclient.example.org%2Fcb    
        &login_hint=tom@example.com
    ```

**Comment by skasund** [comment]
L2 has requested STE slides for this feature.  The STE template can be found at the link below: 

https://ibm.box.com/s/1an42g7zdgmaj84w7dft0indqfgi8ffm

Please upload the completed slides to the same STE Archive BOX folder.  Thanks!

**Comment by marikaj123** [comment]
@ayoho - Do you have English files for this feature that require translation?

**Comment by marikaj123** [comment]
@ayoho Please add the link to the AHA feature.

**Comment by ayoho** [comment]
AHA feature: https://bigblue.aha.io/features/SEC-45

**Comment by skasund** [comment]
L2 has requested STE slides for this feature.  The STE template can be found at the link below: 

https://ibm.box.com/s/1an42g7zdgmaj84w7dft0indqfgi8ffm

Please upload the completed slides to the same STE Archive BOX folder.  Thanks!

**Comment by ayoho** [comment]
@marikaj123 There's one English file with a new message that requires translation:
- dev/com.ibm.ws.security.openidconnect.clients.common/resources/com/ibm/ws/security/openidconnect/common/internal/resources/OidcCommonMessages.nlsprops

```
# 0=Client ID, 1=List of invalid parameter names, 2=forwardLoginParameter
BLACKLISTED_FORWARD_AUTHZ_PARAMS_CONFIGURED=CWWKS1783W: The OpenID Connect client [{0}] includes {1} in the values that are specified by the [{2}] configuration attribute. The specified values will be ignored.
BLACKLISTED_FORWARD_AUTHZ_PARAMS_CONFIGURED.explanation=The OAuth 2.0 specification defines a set of reserved request parameters that must be provided by the client for authorization requests. Values for those reserved parameters cannot be supplied by the user.
BLACKLISTED_FORWARD_AUTHZ_PARAMS_CONFIGURED.useraction=Remove the specified values from the list of values defined in the OpenID Connect client configuration.
```

**Comment by marikaj123** [comment]
Approving since Adam confirmed the translated messages are in the Master branch.

**Comment by marikaj123** [comment]
@ayoho - Please confirm the translated files made it into master.

**Comment by donbourne** [comment]
**Serviceability Approval Comment** - Please answer the following questions for serviceability approval:

1.  WAD -- does the WAD identify the most likely problems customers will see and identify how the feature will enable them to diagnose and solve those problems without resorting to raising a PMR?  Have these issues been addressed in the implementation?

2.  Test and Demo -- As part of the serviceability process we're asking feature teams to test and analyze common problem paths for serviceability and demo those problem paths to someone not involved in the development of the feature (eg. L2, test team, or another development team).  
a) What problem paths were tested and demonstrated? 
b) Who did you demo to?
c) Do the people you demo'd to agree that the serviceability of the demonstrated problem scenarios is sufficient to avoid PMRs for any problems customers are likely to encounter, or that L2 should be able to quickly address those problems without need to engage L3?

3.  SVT -- SVT team is often the first team to try new features and often encounters problems setting up and using them.  Note that we're not expecting SVT to do full serviceability testing -- just to sign-off on the serviceability of the problem paths they encountered.
a) Who conducted SVT tests for this feature?
b) Do they agree that the serviceability of the problems they encountered is sufficient to avoid PMRs, or that L2 should be able to quickly address those problems without need to engage L3?  

4.  Which L2 / L3 queues will handle PMRs for this feature?  Ensure they are present in the contact reference file and in the queue contact summary, and that the respective L2/L3 teams know they are supporting it.  Ask Don Bourne if you need links or more info.

**Comment by gscottj** [comment]
Based on the description provided, it appears that this feature does not have any end-user user interfaces that require accessibility compliance testing.
